include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -228,71 +225,394 @@ struct __merge_sort_leaf_submitter<__internal::__optional_kernel_name<_LeafSortN
+        //                 |  /                        |
+        //                 | /                         |
+        //       i_elem -> +/                          |
+        //                 |                           |
+        //                 |                           |
+        //                 |                           |
+        //                 |                           |
+        //                 |                           |
+        //                 +---------------------------+
+
+        _IndexT i_elem = 0;       // Global diagonal index
+        _IndexT i_elem_local = 0; // Local diagonal index
+        // Offset to the first element in the subrange (i.e. the first element of the first subrange for merge)
+        _IndexT offset = 0;
+        _IndexT n1 = 0; // Size of the first subrange
+        _IndexT n2 = 0; // Size of the second subrange
+
+        WorkDataArea(const std::size_t __n, const std::size_t __n_sorted, const std::size_t __linear_id,
+                     const std::size_t __chunk)
+        {
+            // Calculate global diagonal index
+            i_elem = __linear_id * __chunk;
+
+            // Calculate local diagonal index
+            i_elem_local = i_elem % (__n_sorted * 2);
+
+            // Calculate offset to the first element in the subrange (i.e. the first element of the first subrange for merge)
+            offset = std::min<_IndexT>(i_elem - i_elem_local, __n);
+
+            // Calculate size of the first and the second subranges
+            n1 = std::min<_IndexT>(offset + __n_sorted, __n) - offset;
+            n2 = std::min<_IndexT>(offset + __n_sorted + n1, __n) - (offset + n1);
+        }
+
+        inline bool
+        is_i_elem_local_inside_merge_matrix() const
+        {
+            return i_elem_local < n1 + n2;
+        }
+    };
+
+    template <typename Rng>
+    struct DropViews
+    {
+        using __drop_view_simple_t = oneapi::dpl::__ranges::drop_view_simple<Rng, _IndexT>;
+
+        __drop_view_simple_t rng1;
+        __drop_view_simple_t rng2;
+
+        DropViews(Rng& __rng, const WorkDataArea& __data_area)
+            : rng1(__rng, __data_area.offset), rng2(__rng, __data_area.offset + __data_area.n1)
+        {
+        }
+    };
+
+    std::size_t
+    tune_amount_of_base_diagonals(std::size_t __n_sorted, std::size_t __amount_of_base_diagonals) const
+    {
+        // Multiply work per item by a power of 2 to reach the desired number of iterations.
+        // __dpl_bit_ceil rounds the ratio up to the next power of 2.
+        const std::size_t __k =
+            oneapi::dpl::__internal::__dpl_bit_ceil((std::size_t)std::ceil(256 * 1024 * 1024 / __n_sorted));
+
+        return oneapi::dpl::__internal::__dpl_ceiling_div(__amount_of_base_diagonals, __k);
+    }
+
+    // Calculate nd-range params
+    template <typename _ExecutionPolicy>
+    nd_range_params
+    eval_nd_range_params(_ExecutionPolicy&& __exec, const std::size_t __rng_size) const
```

Comment:
Does it make sense for this to be a constructor of `nd_range_params` rather than a free function?

Reply:
I don't think so, because in the future we may have some different ways of their calculation.

Reply:
OK fair enough.